docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,167 @@
+whose behalf it performs state replication. There are currently three transport
+options available for ABCI applications:
+
+1. **In-process**: Applications written in Go can be linked directly into the
+   same binary as the consensus node. Such applications use a "local" ABCI
+   client, which exposes application methods to the node as direct function
+   calls.
+
+2. **Socket protocol**: Out-of-process applications may export the ABCI service
+   via a custom socket protocol that sends requests and responses over a
+   Unix-domain or TCP socket connection as length-prefixed protocol buffers.
+   In Tendermint, this is handled by the [socket client][socket-client].
+
+3. **gRPC**: Out-of-process applications may export the ABCI service via gRPC.
+   In Tendermint, this is handled by the [gRPC client][grpc-client].
+
+Both the out-of-process options (2) and (3) have a long history in Tendermint.
+The beginnings of the gRPC client were added in May 2016 when ABCI was still
+hosted in a separate repository, and the socket client (formerly called the
+"remote client") was part of ABCI from its inception in November 2015.
+
+At that time when ABCI was first being developed, the gRPC project was very new
+(it launched Q4 2015) and it was not an obvious choice for use in Tendermint.
+It took a while before the language coverage and quality of gRPC reached a
+point where it could be a viable solution for out-of-process applications.  For
+that reason, it made sense for the initial design of ABCI to focus on a custom
+protocol for out-of-process applications.
+
+## Problem Statement
+
+For practical reasons, ABCI needs an interprocess communication option to
+support applications not written in Go. The two practical options are RPC and
+FFI, and for operational reasons an RPC mechanism makes more sense.
+
+The socket protocol has not changed substantially since its original design,
+and has the advantage of being simple to implement in almost any reasonable
+language.  However, its simplicity includes some limitations that have had a
+negative impact on the stability and performance of out-of-process applications
+using it. In particular:
+
+- The protocol lacks request identifiers, so the client and server must
+  complete requests in strict FIFO order. Even if the client issues requests
+  that have no dependency on each other, the protocol has no way except order
+  of issue to map responses to requests. This strictly limits the application's
+  ability to process requests concurrently, and has been a source of complaints
+  from some network operators on that basis.
+
+- The protocol lacks method identifiers, so the only way for the client and
+  server to understand which operation is requested is to dispatch on the type
+  of the request and response payloads. For responses, this means that any
+  error condition is terminal not only to the request, but to the entire ABCI
+  connection.
```

Comment:
Should I understand this to mean that any time the application fails to respond to an ABCI request, the entire socket client breaks? I know that's the case, I'm just confirming that's what's being said here.

Reply:
> Should I understand this to mean that any time the application fails to respond to an ABCI request, the entire socket client breaks? I know that's the case, I'm just confirming that's what's being said here.

I'll clarify the language in the document, but essentially yes: If the server sends back an error message in response to any pending request, the receiver thread in the socket client logs an error and then exits, leaving the client permanently stalled (until the node shuts down). I believe the plumbing outside the client turns these into panics but it would be very easy to disrupt that (say, as part of an ongoing effort to replace panics with error handling).